OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -322,6 +216,174 @@ void TracerThread::Run(
+      close(sample_fd);
+      // Do not remove the ring buffer from fds_to_ring_buffer here as
+      // we are already iterating over fds_to_ring_buffer.
+      fds_to_remove_.push_back(sample_fd);
+      threads_to_fd_.erase(sample_fd);
+    }
+  }
+}
+
+void TracerThread::ProcessMmapEvent(const perf_event_header& header,
+                                    PerfEventRingBuffer& ring_buffer) {
+  // There was a call to mmap with PROT_EXEC, hence refresh the maps.
+  // This should happen rarely.
+  ring_buffer.SkipRecord(header);
+  int fd = ring_buffer.GetFileDescriptor();
+  auto event =
+      std::make_unique<MapsPerfEvent>(MonotonicTimestampNs(), ReadMaps(pid_));
+  event->SetFileDescriptor(fd);
+  DeferEvent(std::move(event));
+}
+
+void TracerThread::ProcessSampleEvent(const perf_event_header& header,
+                                      PerfEventRingBuffer& ring_buffer) {
+  int fd = ring_buffer.GetFileDescriptor();
+  bool is_probe = uprobe_fds_to_function_.count(fd) > 0;
+  bool is_uprobe = is_probe && header.size > 32;  // TODO: find better test.
+  bool is_uretprobe = is_probe && !is_uprobe;
+
+  if (is_uprobe) {
+    auto event = std::make_unique<UprobesWithStackPerfEvent>();
+    ring_buffer.ConsumeRecord(header, &event->ring_buffer_record);
+    event->SetFunction(uprobe_fds_to_function_.at(fd));
+    event->SetFileDescriptor(fd);
+    DeferEvent(std::move(event));
+    ++uprobes_count_;
+  } else if (is_uretprobe) {
+    auto event = std::make_unique<UretprobesPerfEvent>();
+    ring_buffer.ConsumeRecord(header, &event->ring_buffer_record);
+    event->SetFunction(uprobe_fds_to_function_.at(fd));
+    event->SetFileDescriptor(fd);
+    DeferEvent(std::move(event));
+    ++uprobes_count_;
+  } else {
+    auto event = std::make_unique<StackSamplePerfEvent>();
+    ring_buffer.ConsumeRecord(header, &event->ring_buffer_record);
+    event->SetFileDescriptor(fd);
+    DeferEvent(std::move(event));
+    ++sample_count_;
+  }
+}
+
+void TracerThread::ProcessLostEvent(const perf_event_header& header,
+                                    PerfEventRingBuffer& ring_buffer) {
+  LostPerfEvent event;
+  ring_buffer.ConsumeRecord(header, &event.ring_buffer_record);
+  LOG("Lost %lu events", event.GetNumLost());
+}
+
+void TracerThread::DeferEvent(std::unique_ptr<PerfEvent> event) {
+  std::lock_guard<std::mutex> lock(deferred_events_mutex_);
+  deferred_events_.emplace_back(std::move(event));
+}
+
+std::vector<std::unique_ptr<PerfEvent>> TracerThread::ConsumeDeferredEvents() {
+  std::lock_guard<std::mutex> lock(deferred_events_mutex_);
+  std::vector<std::unique_ptr<PerfEvent>> events(std::move(deferred_events_));
+  deferred_events_.clear();
+  return events;
+}
+
+void TracerThread::ProcessDeferredEvents(
+    const std::shared_ptr<std::atomic<bool>>& exit_requested) {
+  while (!(*exit_requested)) {
+    std::vector<std::unique_ptr<PerfEvent>> events = ConsumeDeferredEvents();
+    if (events.size() == 0) {
+      usleep(1'000);
```

Comment:
This wait here looks suspicious. Does this have to be the same 1'000 as in the other thread? I suspect no, but given that they are both 1'000, this may be misunderstood, so I'd suggest defining both as different constants.

Also wondering if this should rather use a wait/notify mechanism rather than the check/sleep.